p2panda-rs/src/document/document_view_id.rs

```diff
@@ -34,13 +34,33 @@ pub struct DocumentViewId(Vec<OperationId>);
 impl DocumentViewId {
     /// Create a new document view id.
     pub fn new(graph_tips: Vec<OperationId>) -> Self {
-        Self(graph_tips)
+        let mut graph_tips_mut = graph_tips;
+        graph_tips_mut.sort();
+        Self(graph_tips_mut)
     }
 
     /// Get the graph tip ids of this view id.
     pub fn graph_tips(&self) -> &[OperationId] {
         self.0.as_slice()
     }
+
+    /// Returns a hash over the graph tips constituting this view id.
+    ///
+    /// Use this as a unique identifier for a document if you need a value with a limited size. The
+    /// document view id itself grows with the number of graph tips that the document has, which
+    /// may not be desirable for an identifier.
+    ///
+    /// Keep in mind that when you refer to document views with this hash value it will not be
+    /// possible to recover the document view id from it.
+    pub fn hash(&self) -> Hash {
+        let graph_tip_bytes = self
+            .0
+            .clone()
```

Comment:
I somehow assumed to find `sort` here as the specification reads that before hashing the graph tips need to be sorted, but then I realized you moved it up to make them comparable which also makes sense.

How about updating the specification that *any* document view id needs to hold operation ids which are sorted. Then this is reflected better in the code-base and brings consistency across all data.

We would still need to make sure they are sorted when coming out of deserialized data, since the constructor is *not* used in that case. I can imagine that could lead to bugs when hashing document view ids coming from CBOR.

Slightly related question: How about sorting in `previous_operations`?

It totally makes sense to assure that the graph tips in `DocumentViewId` are always sorted and not for the document view hash only.

Reply:
I've added that test now :+1:

Reply:
Yeh, I also expected the sorting to happen before hashing, rather than in the `DocumentViewId` constructor. What is the case for it happening there? Is there a down side to it happening in the hashing method?

Reply:
1) Major reason: If we don't sort in the constructor, two document view ids are not equal if they are constructed from the same operation ids in a different order. I added an `equality` test for this, that will fail otherwise.
2) Minor reason: If we don't sort in the constructor every call to `hash()` will require sorting the ids again. If we sort in the constructor it only has to happen when the document view id is first created, after that they will never have to be sorted again, even after serialisation, because passing in an already sorted document view id into the constructor will result in `sort()` just passing over the values without doing anything (as they are already sorted).

Reply:
Without a clear idea where this will lead, here two additional thoughts:

Re: 1) We could also implement `Eq` ourselves and sort inside of it. So then we _only_ sort if we want to compare and if we want to hash. I assume both cases are currently not really used (or in other words: We currently don't need document view hashes).

Re: 2) Graph tip arrays will always be fairly small (1-5 items) so I don't think that sorting is such a heavy penalty. Hashing in itself is probably the more expensive operation here.

Reply:
If we sort in the constructor we need an idea how we sort for deserialization, if we sort in `hash` and `Eq` we don't.. The latter also maybe gives us a peace of mind in regards of enforcing sorting in the CBOR data as I somehow wonder if this is _really_ required?

Reply:
Other thought: Whats more expensive, sorting whenever you use something or always checking if something _is_ sorted? :thinking:

Reply:
Thanks for the explanations @cafca. For point 1) looking at this from the functionality of a DocumentViewId I would expect two IDs to be equal even if the hashes are listed in different orders. The order has no meaning wrt which document view you are refering too. So the custom eq approach does make sense to me.

I agree we should avoid unnecessary sorting, but I wonder if we will be constructing IDs more often than we're hashing them?

Reply:
Ayayayay. 🤣 Now after this discussion I feel like we should implement `Eq` ourselves where we sort before, add sorting into `hash` as well while removing it from the constructor.

It seems to be the best approach plus we don't need to worry about validation too much!

Reply:
Sorting is really inexpensive if the data is already sorted so I think it won't be a problem at all to have this in the constructor. And then we can just derive `Eq` without a custom implementtation.

Reply:
> Sorting is really inexpensive if the data is already sorted so I think it won't be a problem at all to have this in the constructor. And then we can just derive `Eq` without a custom implementtation.

Thats also alright, then we sort everywhere all the time 🤣

I'd be fine with this, as long as we remove the validation again 👍

Reply:
Yep, cool, I'm fine with this too. It doesn't need to be in the spec that `DocumentViewId` must contain sorted hashes, right? It's just that we do it in advance as it _is_ required for hashing them.

@adzialocha what do you mean by removing the validation?

Reply:
During our discussion we've decided to add the `Validate` trait to `DocumentViewId` and that it _needs_ to be sorted. Also there was a discussion in the chat about that topic and we've been reasoning about that any array of document view ids need to be sorted in the CBOR.

Reply:
https://github.com/p2panda/p2panda/pull/277/files#diff-1abf4b4e983cf2c508fc8c65ce494e69129778eca08896b7ef468ed9b03c829aR71-R79

Reply:
If we don't have the requirement that document view ids are sorted we can't derive `PartialEq` because it would think that two instances with differently sorted but same operation ids inside are not the same.

Then we have to implement `PartialEq` ourselves and in that implementation sorts everytime that equality of some document view ids is checked. We would also have to sort everytime `hash()` is accessed.

The alternative is to sort the ids once when they are first created and add a requirement to the spec that document view ids must always be sorted. We have a `sort` call in the constructor then but it won't do much sorting because calling `sort()` on already sorted values is inexpensive.

Reply:
It just feels like we're introducing a requirement of `DocumentViewId` into the spec which is actually a requirement of `DocumentViewHash` (which are the hash or _sorted_ document view ids). That's my only query if you both think this is ok, then lets press on :+1: I'm happy with any of the implementation solutions.

Reply:
It's not just the hash, also checking equality of a set in general is more cumbersome if you don't have sorted data to begin with. I think we should have a requirement for sorted values in all places where we have sets of hashes, unless their order has some special significance.

Reply:
Rejecting unsorted ids seems like quite unexpected behaviour to me. The hash is a product of the id, so I feel it's more intuitive for the sorting to be _enforced_ at the point of hashing. I see the case for sorting in the constructor of `DocumentViewId` in our implementation though.

Reply:
> It's not just the hash, also checking equality of a set in general is more cumbersome if you don't have sorted data to begin with. I think we should have a requirement for sorted values in all places where we have sets of hashes, unless their order has some special significance.

Ok, if it's a more general thing we want in the spec, that makes more sense to me.

Reply:
Maybe we don't need to tie this PR to a spec level decision about general sorting of lists in this PR though?

Reply:
It's gonna rain PRs! But yes, I can split that up as well.

Reply:
Haha, I am ready to get wet 🤣

Reply:
(I forgot my umbrella)